### PR TITLE
clean up disconnect, cancel handlers

### PR DESCRIPTION
**Pull request by garlick**
This is a bit of cleanup peeled off of other work.   

A common pattern when implementing a flux service is to keep a list of pending (often streaming) requests, and then implement disconnect and cancel request handlers that can remove items from this list.   This code is repeated, with differing implementations, throughout the code base.

This PR adds some reusable code, with unit tests, to assist:
* a function to match "sender" route frames without allocating memory
* a list container for messages that uses message reference counts and is safe for item removal during iteration
* helpers for disconnect and cancel requests

The list container is actually an update of an existing one, `libutil/msglist.c`, which is used as a message queue in the handle implementation and connectors.  The update switches it to zlistx_t internally, and hard-wires the duplicator and destructor.  The new version is made part of the Flux API.

I converted a few disconnect/cancel handlers but some were actually quite complex, so I stopped short of converting everything in this PR.

One observation:  we commonly keep a list of requests as a list of structs containing the request plus some state.  If we store the state as an aux item in the message, then in addition to being able to use the above container and helpers, the request state destruction is conveniently tied to message destruction.  I think that's a good pattern to adopt, but converting across the boards will require a bit of work.

### CODE REVIEW

**Review by grondo — APPROVED** [review]
Looks like great cleanup!

Didn't really see anything worthy of a comment.

**Review by chu11 — COMMENTED** [review]
overall LGTM, one typo nit and maybe a rebase error?  I removed `merge-when-passing` just in case you wanted to fix these.

Also, do we maybe want to start an issue to document other places we can cleanup w/ the discconect/cancel common code?  Off the top of my head kvs-watch and job-list.

File: src/common/libflux/handle.c [hunk 1]
```diff
@@ -746,23 +743,6 @@ static int requeue (flux_t *h, flux_msg_t *msg, int flags)
...
-    }
-
-    if (requeue (h, cpy, flags) < 0) {
-        flux_msg_destroy (cpy);
-        return -1;
-    }
-
-    return 0;
-}
-
```
[the remark this answers is not in the document]
Comment (chu11) [reply src/common/libflux/handle.c:765]: oh wait, nevermind, I see that you renamed `requeue()` to `flux_requeue()`.

**Comment by garlick** [comment]
Thanks - I fixed that commit message typo, rebased on current master, and forced a push.  I'll go ahead and set MWP.

**Comment by garlick** [comment]
The bionic test-install builder seems to have hung in configure, so I  restarted it.

**Pull request merged**